From 7b971e88d3249b89181f084f96849d314a0f522c Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 1 Feb 2020 13:56:53 +0100
Subject: [PATCH 1/1] efi_loader: provide function for arch specific setup

On RISC-V the id of the active HART has to be passed as a UEFI variable
for booting Linux via UEFI.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 include/efi_loader.h       |  3 +++
 lib/efi_loader/efi_setup.c | 14 ++++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/include/efi_loader.h b/include/efi_loader.h
index d4c59b54c4..70b7f09e03 100644
--- a/include/efi_loader.h
+++ b/include/efi_loader.h
@@ -116,6 +116,9 @@ extern efi_uintn_t efi_memory_map_key;
 extern struct efi_runtime_services efi_runtime_services;
 extern struct efi_system_table systab;
 
+/* Architecture specific initialization of the UEFI system */
+efi_status efi_setup_arch_specific(void);
+
 extern struct efi_simple_text_output_protocol efi_con_out;
 extern struct efi_simple_text_input_protocol efi_con_in;
 extern struct efi_console_control_protocol efi_console_control;
diff --git a/lib/efi_loader/efi_setup.c b/lib/efi_loader/efi_setup.c
index de7b616c6d..69d27829e5 100644
--- a/lib/efi_loader/efi_setup.c
+++ b/lib/efi_loader/efi_setup.c
@@ -22,6 +22,16 @@ void __weak allow_unaligned(void)
 {
 }
 
+/**
+ * efi_setup_arch_specific() - architecture specific UEFI setup
+ *
+ * This routine can be used to define architecture specific variables
+ * or configuration tables, e.g. HART id for RISC-V
+ */
+efi_status __weak  efi_setup_arch_specific(void)
+{
+}
+
 /**
  * efi_init_platform_lang() - define supported languages
  *
@@ -178,6 +188,10 @@ efi_status_t efi_init_obj_list(void)
 	ret = efi_reset_system_init();
 	if (ret != EFI_SUCCESS)
 		goto out;
+	
+	ret = efi_setup_arch_specific();
+	if (ret != EFI_SUCCESS)
+		goto out;
 
 out:
 	efi_obj_list_initialized = ret;
-- 
2.24.1

